### PR TITLE
feat: improve signup flow

### PR DESCRIPTION
**Pull request by kittybest**
- [x] lint all files
- [x] move signup button to `ConnectedDetails` (which still display `UserInfo` button to let the user be able to switch account)
- [x] remove `EligibilityDialog` in `ConnectButton` since we are using `useMaciSignup` hook to tell if the user is eligible or not
- [x] make display logic clearer in `BallotOverview` and `AddToBallot` components
- [x] hide `ProjectSelectButton` if the user is not registered

### CODE REVIEW

File: src/components/ConnectButton.tsx [hunk 1]
```diff
@@ -163,3 +162,24 @@ const SignupButton = ({
...
+      size="sm"
+      isOpen={true}
+      onOpenChange={onClose}
+      title={signupError}>
+      <p>
+      {
+        signupError === MaciSignUpError.NO_KEY ? "You don't have a maci key yet. Please reconnect your wallet and try again."
+        : signupError === MaciSignUpError.NO_ATTESTATION ? "You don't have an attestation yet. Try to get one and refresh page."
+        : signupError === MaciSignUpError.NO_SIGNER ? "Something wrong with your wallet. Please reconnect again."
+        : "Unknown error happened."
```
Comment (0xmad) [inline src/components/ConnectButton.tsx:180]: It's not a good practice to use nested `?:` operators.

File: src/hooks/useMaciSignup.ts [hunk 2]
```diff
@@ -8,10 +8,19 @@ import { api } from "~/utils/api";
...
+  NO_SIGNER = "No Signer",
+  UNKNOWN = "Unknown",
+}
+
 export interface IUseMaciSignUpData {
   stateIndex?: string;
   isRegistered?: boolean;
   onSignup: () => Promise<void>;
+  error?: MaciSignUpError;
+  setError: (error?: MaciSignUpError) => void
```
Comment (0xmad) [inline src/hooks/useMaciSignup.ts:23]: Why do you need to expose this state setter?

File: src/hooks/useMaciSignup.ts [hunk 3]
```diff
@@ -8,10 +8,19 @@ import { api } from "~/utils/api";
...
+  NO_ATTESTATION = "Don't Have Attestation",
+  NO_SIGNER = "No Signer",
+  UNKNOWN = "Unknown",
+}
+
 export interface IUseMaciSignUpData {
   stateIndex?: string;
   isRegistered?: boolean;
   onSignup: () => Promise<void>;
+  error?: MaciSignUpError;
```
Comment (0xmad) [inline src/hooks/useMaciSignup.ts:22]:
```suggestion
  stateIndex?: string;
  error?: MaciSignUpError;
  isRegistered?: boolean;
  onSignup: () => Promise<void>;
```

File: src/hooks/useMaciSignup.ts [hunk 4]
```diff
@@ -34,7 +45,16 @@ export const useMaciSignup = (): IUseMaciSignUpData => {
...
+      return;
+    }
+    if (!signer) {
+      setError(MaciSignUpError.NO_SIGNER);
+      return;
+    }
+    if (!attestationId) {
+      setError(MaciSignUpError.NO_ATTESTATION);
       return;
     }
```
Comment (0xmad) [inline src/hooks/useMaciSignup.ts:59]: These cases are not possible to get because you can't call signup if there is no attestation id, signer or public key.

File: src/features/projects/components/AddToBallot.tsx [hunk 5]
```diff
@@ -47,7 +49,7 @@ export const ProjectAddToBallot = ({ id, name }: Props) => {
         </IconButton>
       ) : (
         <Button
-          disabled={!address}
+          disabled={!address || !isRegistered}
```
Comment (0xmad) [inline src/features/projects/components/AddToBallot.tsx:52]: If user is not allowed to vote, this button should be hidden, not disabled.

File: src/features/ballot/components/BallotOverview.tsx [hunk 6]
```diff
@@ -102,7 +105,11 @@ function BallotOverview() {
           </div>
         </div>
       </BallotSection>
-      {ballot?.publishedAt ? (
+      { !isRegistered ? (
+        <Button className={"w-full"} variant="primary" disabled>
+          Sign up to vote
+        </Button>
+      ) : ballot?.publishedAt ? (
```
Comment (0xmad) [inline src/features/ballot/components/BallotOverview.tsx:112]: Please, run prettier for all your changes manually.

File: src/features/ballot/components/BallotOverview.tsx [hunk 7]
```diff
@@ -102,7 +105,11 @@ function BallotOverview() {
           </div>
         </div>
       </BallotSection>
-      {ballot?.publishedAt ? (
+      { !isRegistered ? (
+        <Button className={"w-full"} variant="primary" disabled>
```
Comment (0xmad) [inline src/features/ballot/components/BallotOverview.tsx:109]:
```suggestion
        <Button className="w-full" variant="primary" disabled>
```

Comment (0xmad) [reply src/features/ballot/components/BallotOverview.tsx:109]: Why this button is disabled?

File: src/components/ConnectButton.tsx [hunk 8]
```diff
@@ -68,15 +67,6 @@ export const ConnectButton = () => {
...
               }
 
-              if (!isRegistered) {
-                return (
-                  <SignupButton
-                    loading={isRegistered === undefined}
-                    onClick={onSignup}
-                  />
-                );
-              }
```
Comment (0xmad) [inline src/components/ConnectButton.tsx:78]: Why did you remove it?

**Review by 0xmad — CHANGES_REQUESTED** [review]
@kittybest thanks, just few comments

File: src/hooks/useMaciSignup.ts (hunk 2 above)
Comment (kittybest) [reply src/hooks/useMaciSignup.ts:23]: actually I just wanna call **set error to `undefined`**, I should add a function instead of exposing the state setter :)

File: src/hooks/useMaciSignup.ts (hunk 4 above)
Comment (kittybest) [reply src/hooks/useMaciSignup.ts:59]: so it will look like this:

https://github.com/privacy-scaling-explorations/maci-rpgf/assets/9023842/3bad6765-de07-4bc8-ada0-7acb80172ec5

Comment (0xmad) [reply src/hooks/useMaciSignup.ts:59]: If you don't have attestation signup button won't be shown, the same with signer if you don't connect the wallet, and with public key because it's created for the session.

Comment (0xmad) [reply src/hooks/useMaciSignup.ts:59]: Showing modal for errors is also a bad ux pattern.

File: src/features/projects/components/AddToBallot.tsx (hunk 5 above)
Comment (0xmad) [reply src/features/projects/components/AddToBallot.tsx:52]: Take a look at my PR #21. It's implemented there.

File: src/components/ConnectButton.tsx (hunk 8 above)
Comment (kittybest) [reply src/components/ConnectButton.tsx:78]: because I move it to `ConnectedDetails`.
why doing this is because currently, if I connect my wallet, and I don't have attestation for my current address, I have no way to disconnect and reconnect another wallet address. 
which means, I wanna display two buttons on the navbar once the user connected their wallet -- one is the signup button, the other one is the user info button, which allow the user to disconnect their wallet.

File: src/features/projects/components/AddToBallot.tsx (hunk 5 above)
Comment (kittybest) [reply src/features/projects/components/AddToBallot.tsx:52]: I found it, so what should I do now?

**Comment by kittybest** [comment]
@0xmad how do you think about error handling during signup? if not using popup modals, I have no idea how to display them. (e.g. transaction failed)

**Comment by 0xmad** [comment]
> @0xmad how do you think about error handling during signup? if not using popup modals, I have no idea how to display them. (e.g. transaction failed)

Popup is not the same as modal, and demo already has error display logic with popup. So you can reuse it.

**Comment by kittybest** [comment]
> > @0xmad how do you think about error handling during signup? if not using popup modals, I have no idea how to display them. (e.g. transaction failed)
> 
> Popup is not the same as modal, and demo already has error display logic with popup. So you can reuse it.

sorry for the misunderstanding wording 0.0
will reuse that.

File: pnpm-lock.yaml
Comment (0xmad) [inline pnpm-lock.yaml:1]: I don't think we need to change this file. There are no dependency updates.

File: src/components/ConnectButton.tsx [hunk 9]
```diff
@@ -68,19 +66,6 @@ export const ConnectButton = () => {
                 return <Chip onClick={openChainModal}>Wrong network</Chip>;
               }
 
-              if (!isAllowedToVote) {
-                return <Chip>You are not allowed to vote</Chip>;
-              }
```
Comment (0xmad) [inline src/components/ConnectButton.tsx:73]: This is missing part in connect details.

**Comment by kittybest** [comment]
@0xmad I would like to move the error handling in another PR, so this PR only do some minor button display improvements.

File: pnpm-lock.yaml
Comment (0xmad) [inline pnpm-lock.yaml:1]: Seems it still has changes

Comment (kittybest) [reply pnpm-lock.yaml:1]: Just lint

**Review by 0xmad — APPROVED** [review]
@kittybest thanks!
